### PR TITLE
try to solve remaining acl problems - initial commit

### PR DESCRIPTION
**Pull request by andrepereiradasilva**
Pull Request for Several ACL Issues (part 2).
#### Summary of Changes

What this PR does:
- Further simplifies the ACL rules code
- Displays the correct calculated permissions in article <- category <- com_content <- global config scenario
- Stop explicit writing the parent asset permissions on creating a new item (ex: module)
- Solves some other minor bugs
  ...
##### Pre-requisites

Use latest staging.
Clean install.
##### Tests

1: This tests Global config child groups Inherit
- Go to Global config
- Change Public "Delete" to Denied
- Change Public "Edit" to Allowed
- Refresh the page.
- Change Registered "Edit state" to Denied
- Change Registered "Edit Own" to Allowed
- Refresh the page
- Check all Global config Permissions are all inhrited as they should.

2: This tests Global config -> Component Inherit
- Go to Articles (com_content) Config 
- Check the com_config permission are all inhrited as they should.
- Change Author "Edit Own" to Denied
- Refresh the page
- Check the child groups permissions are all inhrited as they should.

3: This tests Global config -> Component -> Item Inherit
- Create new category, name it "Test ACL" don't change anything and save
- Now edit the category
- Check the com_config + com_content permissions are all inhrited as they should.
- Change Editor "Edit" to Denied
- Refresh the page
- Check the child groups  permissions are all inhrited as they should.

4: This tests Global config -> Component -> Extension -> Item Inherit
- Create a new article assign to "Test ACL" category, don't change anything and save
- Now edit the article
- Check the com_config + com_content + category permissions are all inhrited as they should.
- Change Manager "Edit" to Denied
- Refresh the page
- Check the child groups  permissions are all inhrited as they should.

5: This tests creating anew item without explicit inhreting permissions
- Create a new article assign to "Test ACL 2" category, don't change anything and save
- Now edit the article
- Check the com_config + com_content + category permissions are all inhrited as they should.
- Check there is no explict "Allowed" or "Denied" permission.
#### Know issues for other(s) PR

1: If a component as a permission that doesn't exists in global config (ex: frontend editing in com_modules) we get "Not Allowed (Inherited)" when we should get "Not Allowed (Default)"

2: When creating a new item (not saving) if uses the calculated permissions from the component (ex: module <- com_modules <- global config), but not if you have a section too (ex: article <- category <- com_content <- global config)

3: In"2:", it uses the component permission, but should use the child of componet calculated permissions.

4: When changing a permission of an item that doesn't have a row in the asset table the row is created fine for module <- com_modules <- global config scenario, but not for article <- category <- com_content <- global config scenario.

### CODE REVIEW

**Comment by infograf768** [comment]
Tested here on a User child of administrator and articles.
Component settings:
screen shot 2016-06-20 at 11 12 25

Before patch for New article
screen shot 2016-06-20 at 11 13 03

After patch for New article
screen shot 2016-06-20 at 11 17 29

Tested also with Modules and our Content Languages new ACL.
Works OK for me (no code review).

**Comment by infograf768** [comment]
For menus, it looks like it is inconsistent 

Component Options:

screen shot 2016-06-20 at 11 49 38

For a specific menu

screen shot 2016-06-20 at 11 51 06

But for another, as well as for a new menu

screen shot 2016-06-20 at 11 52 20

For both these menus the parent group has same permissions, i.e

screen shot 2016-06-20 at 11 53 51

**Comment by infograf768** [comment]
And I see that when editing an existing module I also get Locked instead of inherited and can't modify.

**Comment by andrepereiradasilva** [comment]
in reality, IMO, the permission tab shouldn't even exist when you are creating a new item

**Comment by andrepereiradasilva** [comment]
Anyway i will check the issues you described.

**Comment by andrepereiradasilva** [comment]
@infograf768 , reggarding your comments you must not forget that an action/group in a item inherit from:
- parent group(s) in the same asset, if any
- same group in parent assets (category -> component -> global config), if any

So for instance, a explicit "Denied" in "Public" global config will be a "Not Allowed (Locked)" across all joomla user groups, components, categories and items for "Public" and their child groups.

File: administrator/components/com_config/model/application.php [hunk 1]
```diff
@@ -577,6 +582,21 @@ public function storePermissions($permission = null)
...
 			$assetId = (int) $this->db->loadResult();
 
+			// Fetch the parent asset id.
+			$parentAssetId = null;
+
+			// If not in global config we need the parent_id asset to calculate permissions.
+			if (!$isGlobalConfig)
+			{
+				// In this case we need to get the component rules too.
+				$query = $this->db->getQuery(true)
```
Comment (roland-d) [inline administrator/components/com_config/model/application.php]: You can use

```
$query->clear()
```

as $query is already instantiated.

**Comment by roland-d** [comment]
@andrepereiradasilva I have tested all 5 scenarios successful. I do agree we should not show the permission tab on creating a new article or at least make it blurred.

**Comment by roland-d** [comment]
@andrepereiradasilva Sorry, previous post removed because I realized I tested without patch applied ;)

Test 1: :+1: 
Test 2: :+1: 
Test 3: :+1:
Test 4: :+1:
Test 5: :+1:

**Comment by infograf768** [comment]
Ok, I think I found where the issues came from here
Example for menus: I had already played with permissions for some of them before the ACL patch (and maybe before you corrected some issues with the new menus ACL, therefore I got weird results _THAT could not be changed though by saving again the menu_ I had to delte the permissions in the db for these menus to get the correct ones (locked).

Example: A menu which had already permissions:
screen shot 2016-06-21 at 07 46 51

A menu which did not have permissions:
screen shot 2016-06-21 at 07 48 01

So, the problem is that it is not B/C for some menus here (but menus ACL is new to 3.6.0).

For Modules:
general Options (I get unresponsive script btw and have to click Continue as, I guess, I have many modules):

screen shot 2016-06-21 at 08 22 07

For any single module ( never touched the permissions on that test site)
screen shot 2016-06-21 at 08 25 19

But when I save the module, I get the correct

screen shot 2016-06-21 at 08 29 12

So it looks OK as menus ACL are new to 3.6.0
Please propose patch to Staging.

---

Now, I found another bug...
When a group is deleted, the assets table keeps the assets defined for that group...

**Pull request closed without merging**